File: Projects/codechef/BSTOPS.java
package codechef;

import java.io.IOException;
import java.util.Scanner;

class BSTOPS {
    private static Node insert(Node root, int x, long p) {
        if (root == null) {
            System.out.println(p);
            return new Node(x, p);
        } else if (x < root.x)
            root.lt = insert(root.lt, x, p * 2);
        else if (x > root.x)
            root.rt = insert(root.rt, x, p * 2 + 1);
        return root;
    }

    private static Node delete(Node root, int x) {
        if (x < root.x)
            root.lt = delete(root.lt, x);
        else if (x > root.x)
            root.rt = delete(root.rt, x);
        else {
            System.out.println(root.p);
            if (root.lt == null)
                return root.rt;
            else if (root.rt == null)
                return root.lt;
            else {
                Node successor = root.rt;
                while (successor.lt != null)
                    successor = successor.lt;
                root.x = successor.x;
                root.rt = delete(root.rt, successor.x);
            }
        }
        return root;
    }

    public static void main(String[] args) throws IOException {
        Scanner stdin = new Scanner(System.in);
        int n = stdin.nextInt();
        Node root = null;
        while (n-- > 0) {
            char c = stdin.next().charAt(0);
            int x = stdin.nextInt();
            switch (c) {
                case 'i':
                    root = insert(root, x, 1);
                    break;
                case 'd':
                    root = delete(root, x);
                    break;
            }
        }
    }

    private static class Node {
        Node lt, rt;
        int x;
        long p;

        Node(int x, long p) {
            this.x = x;
            this.p = p;
        }
    }
}

File: Projects/codechef/BSTOPS.kt
package codechef

import java.io.IOException
import java.util.*

internal object BSTOPS {
    private fun insert(root: Node?, x: Int, p: Long): Node {
        if (root == null) {
            println(p)
            return Node(x, p)
        } else if (x < root.x) {
            root.lt = insert(root.lt, x, p * 2)
        } else if (x > root.x) {
            root.rt = insert(root.rt, x, p * 2 + 1)
        }
        return root
    }

    private fun delete(root: Node?, x: Int): Node? {
        if (x < root!!.x) {
            root.lt = delete(root.lt, x)
        } else if (x > root.x) {
            root.rt = delete(root.rt, x)
        } else {
            println(root.p)
            if (root.lt == null) {
                return root.rt
            } else if (root.rt == null) {
                return root.lt
            } else {
                var successor = root.rt
                while (successor!!.lt != null) successor = successor.lt
                root.x = successor.x
                root.rt = delete(root.rt, successor.x)
            }
        }
        return root
    }

    @Throws(IOException::class)
    @JvmStatic
    fun main(args: Array<String>) {
        val stdin = Scanner(System.`in`)
        var n = stdin.nextInt()
        var root: Node? = null
        while (n-- > 0) {
            val c = stdin.next()[0]
            val x = stdin.nextInt()
            when (c) {
                'i' -> root = insert(root, x, 1)
                'd' -> root = delete(root, x)
            }
        }
        stdin.close()
    }

    private class Node internal constructor(var x: Int, var p: Long) {
        var lt: Node? = null
        var rt: Node? = null

    }
}